### PR TITLE
Single-pass Self Parameter Tuning (SSPT)

### PR DESCRIPTION
**Pull request by smastelini**
Here comes one more candidate to the estimator arena!

This PR introduces the algorithm presented in the following paper:

`Veloso, B., Gama, J., Malheiro, B., & Vinagre, J. (2021). Hyperparameter self-tuning for data streams. Information Fusion, 76, 75-86.`

I am working with Bruno Veloso (@BrunoMVeloso), the main author of the technique.

**What does SSPT do?**

- Given an estimator and a performance metric, SSPT uses the Nelder-Mead algorithm to "walk" the hyperparameter space in search of the best configuration.

**So, hyperparameter tuning, but what changes?**

- SSPT leverages the internal states of the best models. In other words, it is not a "tournament" where bad estimators are discarded, and new ones are created from scratch. SSPT uses the `mutate` capabilities of River's estimators to keep evolving! The best solutions produce new estimators that share their ancestors' internal states (weights, for instance).

**What is currently supported  (hopefully)?**

- Classifiers, regressors, anomaly detectors, and anomaly filters. I'm using a lot of meta-programming to achieve that :P, but hopefully, it will work. @MaxHalford, when you have some spare time, some tips would be welcome!

**What kind of hyperparameters are supported?**

- Due to the nature of the Nelder-Mead algorithm, only numerical hyperparameters are supported (floats and integers).
 
**Disclaimers:**

1. WIP. I am sure there is much room for improvement regarding code style and computation performance. I'll work with @BrunoMVeloso to check the code for bugs.
2. We can also improve the API. Currently, we have to define the data type of hyperparameters explicitly. We could potentially infer the types via typing annotations, but not everything is "typed" right now.

---

Here's a little bit of a demo of the current capabilities:

```python
from river import datasets
from river import linear_model
from river import metrics
from river import preprocessing
from river_extra import model_selection

dataset = datasets.synth.Friedman(seed=42).take(5000)

sspt = model_selection.SSPT(
    estimator=preprocessing.AdaptiveStandardScaler() | linear_model.LinearRegression(),
    metric=metrics.RMSE(),
    grace_period=500,
    params_range={
        "AdaptiveStandardScaler": {
            "alpha": (float, (0.1, 0.9))
        },
        "LinearRegression": {
            "l2": (float, (0.0, 0.2)),
            "optimizer": {
                "lr": {"learning_rate": (float, (0.0091, 0.5))}
            },
            "intercept_lr": {"learning_rate": (float, (0.0001, 0.5))}
        }
    },
    start="random",
    drift_input=lambda yt, yp: abs(yt - yp),
    convergence_sphere=0.00001,
    seed=42
)
metric = metrics.RMSE()
first_print = True

for i, (x, y) in enumerate(dataset):
    y_pred = sspt.predict_one(x)
    metric.update(y, y_pred)
    sspt.learn_one(x, y)

    if sspt.converged and first_print:
        print("Converged at:", i)
        first_print = False

print("Total instances:", i + 1)
print(metric)
print("Best params:")
print(repr(sspt.best))

```

Which gave me:

```
otal instances: 5000
RMSE: 4.738158
Best params:
Pipeline (
  AdaptiveStandardScaler (
    alpha=0.622852
  ),
  LinearRegression (
    optimizer=SGD (
      lr=Constant (
        learning_rate=0.0091
      )
    )
    loss=Squared ()
    l2=0.
    l1=0.
    intercept_init=0.
    intercept_lr=Constant (
      learning_rate=0.0001
    )
    clip_gradient=1e+12
    initializer=Zeros ()
  )
)
```

I am not sure it is working, TBH 😄, but it didn't crash, at least (as many times before)!

Now, I'll work with @BrunoMVeloso to check if I got the actual hyperparameter optimization part right. I spent most of the time dealing with hyperparameter generation and formatting the param info to be compatible with `clone` and `mutate`. I am sure there is a lot of duplicated logic.

### CODE REVIEW

**Review by BrunoMVeloso — APPROVED** [review]

**Comment by smastelini** [comment]
Some updates after some apparent hiatus:

- Bruno Veloso and his research team have been working on testing and improving SSPT this whole time.

- The current version has some caveats but is functional and consistent with the original paper. Some of the current limitations:
    - Numpy usage. It is still not clear whether or not we can get rid of numpy in some portions of the code
    - Hacky pipeline traversal. SSPT has to check every component in the pipeline hierarchy, and there might be better ways to do that in comparison with the code I wrote last year.
    - Hacky unified API design (to cover multiple tasks. This definitely needs more testing)
    - Better starting strategies and stopping criteria, as well as optimization procedures. Here we enter the research realm. @BrunoMVeloso and his collaborators are working on that. The core SSPT is there; let us see what the future holds :D

To avoid having a huge commit history, I am going to merge this PR for now. 

It is still strictly experimental and new waves of improvement will come.
